Add tutorial for port-Hamiltonian systems (#2132)
pymor/pymor: docs/source/tutorial_ph.md

```

# Tutorial: Model order reduction for port-Hamiltonian systems

## Port-Hamiltonian LTI systems

```{math}
E \dot{x}(t) & = (J - R) Q x(t) + (G-P) u(t), \\
y(t) & = (G+P)^T Q x(t) + (S-N) u(t),
```

where {math}`H := Q^T E`,

```{math}
\Gamma =
\begin{bmatrix}
    J & G \\
    -G^T & N
\end{bmatrix},
\text{ and }
W =
\begin{bmatrix}
    R & P \\
    P^T & S
\end{bmatrix}
```

satisfy
{math}`H = H^T \succ 0`,
{math}`\Gamma^T = -\Gamma`, and
{math}`W = W^T \succcurlyeq 0`.

A dynamical system of this form, together with a given quadratic (energy)
function {math}`\mathcal{H}(x) := \tfrac{1}{2} x^T H x`, typically
called Hamiltonian, is called a port-Hamiltonian system.

In pyMOR, there exists {{ PHLTIModel }}. As of now, pyMOR only supports
port-Hamiltonian systems with nonsingular E. {{ PHLTIModel }} inherits from
{{ LTIModel }}, so {{ PHLTIModel }} can be used with all reductors which expect
a {{ LTIModel }}. However, only specific reductors preserve the
port-Hamiltonian structure, i.e., return a ROM of type {{ PHLTIModel }} or
allow the conversion of the ROM to {{ PHLTIModel }}. In applications, it is
often desirable to preserve the port-Hamiltonian structure.

> It is known that if the LTI system is minimal and stable, the following are equivalent:
> - The system is passive.
> - The system is port-Hamiltonian.
> - The system is positive real.

Reviewer: ```suggestion
It is known that if the LTI system is minimal and stable, the following are equivalent:

- The system is passive.
- The system is port-Hamiltonian.
- The system is positive real.
```

A passive {{ LTIModel }} can be converted into a {{ PHLTIModel }} using
the {func}`~pymor.models.iosys.PHLTIModel.from_passive_LTIModel` function if desired.
Consequentely, one option to preserve port-Hamiltonian structure is to use a reductor
which preserves passivity (but returns a ROM of type {{ LTIModel }}), and convert the
ROM into a {{ PHLTIModel }} in a post-processing step.

## A toy problem: Mass-spring-damper chain

As a toy problem, we use a mass-spring-damper chain, which can be formulated
as a port-Hamiltonian system (see {cite}`GPBV12`):

```{image} msd_example.svg
:alt: MSD example
:width: 60%
```

Here, the spring constants are denoted by {math}`k_i` and the damping constants by
{math}`c_i`, {math}`i=1,\dots,n/2`.
The inputs {math}`u_1` and {math}`u_2` are the external forces on the first two
masses {math}`m_1` and {math}`m_2`. The system outputs {math}`y_1` and {math}`y_2`
correspond to the velocities of the first two masses {math}`m_1` and {math}`m_2`.
The toy problem is included in pyMOR in the {mod}`pymor.models.examples` module as
{func}`~pymor.models.examples.msd_example`.

## Structure-preserving model order reduction

> pyMOR provides three reductors which can be used for model order reduction 

Reviewer: ```suggestion
pyMOR provides three reductors which can be used for model order reduction
```

while preserving the port-Hamiltonian structure:

> - pHIRKA ({class}`~pymor.reductors.ph.ph_irka.PHIRKAReductor`),

Reviewer: ```suggestion
- pH-IRKA ({class}`~pymor.reductors.ph.ph_irka.PHIRKAReductor`),
```

I think the "official" name (from the [paper](https://doi.org/10.1016/j.automatica.2012.05.052)) is IRKA-PH, but I guess PH-IRKA or pH-IRKA is also fine.

- PRBT ({class}`~pymor.reductors.bt.PRBTReductor`),
- passivity preserving model reduction via spectral factorization
  ({class}`~pymor.reductors.spectral_factor.SpectralFactorReductor`).

In this section, we apply all three reductors on our toy example and compare
their performance. All three reductors are described in {cite}`BU22` in more detail.

Note: Currently, the {class}`~pymor.reductors.bt.PRBTReductor` and
{class}`~pymor.reductors.spectral_factor.SpectralFactorReductor` reductors require
{math}`D` to be nonsingular. The MSD example has a zero {math}`D` matrix. Therefore,
we have to add a small regularization feedthrough term. This is a limitation of the

> current implementation, since in the background the so-called KYP-LMI is converted
> into a so-called Riccati equation, which is only possible for nonsingular {math}`D`.

Reviewer: ```suggestion
current implementation, since in the background the KYP-LMI is converted
into a Riccati equation, which is only possible for nonsingular {math}`D`.
```

# by introducing a regularization feedthrough term D
# (required for PRBTReductor and SpectralFactorReductor reductors)
S += np.eye(S.shape[0]) * 1e-12

fom = PHLTIModel.from_matrices(J, R, G, S=S, Q=Q, solver_options={'ricc_pos_lrcf': 'scipy'})
```

The `ricc_pos_lrcf` solver option refers to the solver used for the underlying
Riccati equation relevant for {class}`~pymor.reductors.bt.PRBTReductor` and
{class}`~pymor.reductors.spectral_factor.SpectralFactorReductor`. Possible choices are
`scipy`, `slycot` or `pymess` (if installed).

### pHIRKA

The pHIRKA reductor {class}`~pymor.reductors.ph.ph_irka.PHIRKAReductor` directly returns
a ROM of type {{ PHLTIModel }}. pHIRKA works similar to the general IRKA reductor
{class}`~pymor.reductors.h2.IRKAReductor`, but with fewer degrees of freedom to preserve
the port-Hamiltonian structure.

```{code-cell}
from pymor.reductors.ph.ph_irka import PHIRKAReductor

reductor = PHIRKAReductor(fom)
rom1 = reductor.reduce(10)
print(f'rom1 is of type {type(rom1)}.')
```

### Positive-real balanced truncation (PRBT)

Positive-real balanced truncation (PRBT) works similar to the general balanced truncation

> method described in {doc}`tutorial_bt`, but uses so-called positive real controllability

Reviewer: ```suggestion
method described in {doc}`tutorial_bt`, but uses positive real controllability
```

and observability Gramians instead. PRBT preserves passivity, but returns a ROM
of type {{ LTIModel }}. Thus, we convert the ROM into a {{ PHLTIModel }} in a
post-processing step. PRBT can be used with any passive {{ LTIModel}} FOM.

```{code-cell}
from pymor.reductors.bt import PRBTReductor

reductor = PRBTReductor(fom)

> rom2 = reductor.reduce(10)
> rom2 = PHLTIModel.from_passive_LTIModel(rom2)

Reviewer: ```suggestion
rom2 = reductor.reduce(10)
rom2 = rom2.with_(solver_options={'ricc_pos_lrcf': 'slycot'})
rom2 = PHLTIModel.from_passive_LTIModel(rom2)
```

Reviewer: The solver from `pymess` raises `NotImplementedError` (default when `pymess` is installed, as is the case for the docs building on CI) and `scipy` raises `LinAlgError: The associated Hamiltonian pencil has eigenvalues too close to the imaginary axis`. The `slycot` solver seems to work fine.

> print(f'rom2 is of type {type(rom1)}.')

Reviewer: ```suggestion
print(f'rom2 is of type {type(rom2)}.')
```

```

### Passivity preserving model reduction via spectral factorization

The {class}`~pymor.reductors.spectral_factor.SpectralFactorReductor` method

> is a wrapper reductor for another generic reductor. The method extracts a so-called

Reviewer: ```suggestion
is a wrapper reductor for another generic reductor. The method extracts a
```

spectral factor from the FOM, which subsequentely is reduced by a reductor
specified by the user. A spectral factor is a standard {{ LTIModel }}. Here, we opt
for {class}`~pymor.reductors.h2.IRKAReductor` as the inner reductor. If the inner
reductor returns a stable ROM, passivity is preserved.
The spectral factor method and PRBT are related, since the computation of
the optimal spectral factor for model reduction depends on the computation of
the positive real observability Gramian. The spectral factor method can be used with
any passive {{ LTIModel}} FOM. Again, we convert the ROM of type {{ LTIModel}} into a
{{ PHLTIModel }} in a post-processing step.

```{code-cell}
from pymor.reductors.spectral_factor import SpectralFactorReductor
from pymor.reductors.h2 import IRKAReductor

reductor = SpectralFactorReductor(fom)

> rom3 = reductor.reduce(
>     lambda spectral_factor, mu : IRKAReductor(spectral_factor, mu).reduce(10)
> )
> rom3 = PHLTIModel.from_passive_LTIModel(rom3)

Reviewer: ```suggestion
rom3 = reductor.reduce(
    lambda spectral_factor, mu : IRKAReductor(spectral_factor, mu).reduce(10)
)
rom3 = rom3.with_(solver_options={'ricc_pos_lrcf': 'slycot'})
rom3 = PHLTIModel.from_passive_LTIModel(rom3)
```

> print(f'rom3 is of type {type(rom1)}.')

Reviewer: ```suggestion
print(f'rom3 is of type {type(rom3)}.')
```

```

### Comparison

> Let us compare the H2 errors of the three methods:

Reviewer: ```suggestion
Let us compare the {math}`\mathcal{H}_2` errors of the three methods:
```

      jupytext_version: 1.11.2
kernelspec:
  display_name: Python 3
  name: python3
---

```{try_on_binder}
```

```{code-cell}
:load: myst_code_init.py
:tags: [remove-cell]

```

# Tutorial: Model order reduction for port-Hamiltonian systems

## Port-Hamiltonian LTI systems

```{math}
E \dot{x}(t) & = (J - R) Q x(t) + (G-P) u(t), \\
y(t) & = (G+P)^T Q x(t) + (S-N) u(t),
```

where {math}`H := Q^T E`,

```{math}
\Gamma =
\begin{bmatrix}
    J & G \\
    -G^T & N
\end{bmatrix},
\text{ and }
W =
\begin{bmatrix}
    R & P \\
    P^T & S
\end{bmatrix}
```

satisfy
{math}`H = H^T \succ 0`,
{math}`\Gamma^T = -\Gamma`, and
{math}`W = W^T \succcurlyeq 0`.

A dynamical system of this form, together with a given quadratic (energy)
function {math}`\mathcal{H}(x) := \tfrac{1}{2} x^T H x`, typically
called Hamiltonian, is called a port-Hamiltonian system.

In pyMOR, there exists {{ PHLTIModel }}. As of now, pyMOR only supports
port-Hamiltonian systems with nonsingular E. {{ PHLTIModel }} inherits from
{{ LTIModel }}, so {{ PHLTIModel }} can be used with all reductors which expect
a {{ LTIModel }}.

## MSD problem

```{code-cell}
import numpy as np
from pymor.models.iosys import PHLTIModel

> def msd(n=6, m=2, m_i=4, k_i=4, c_i=1, as_lti=False):

Reviewer: This could be moved to `pymor.models.examples`.

{func}`~pymor.models.examples.msd_example`.

## Structure-preserving model order reduction

pyMOR provides three reductors which can be used for model order reduction
while preserving the port-Hamiltonian structure:

- pH-IRKA ({class}`~pymor.reductors.ph.ph_irka.PHIRKAReductor`) {cite}`GPBV12`,
- PRBT ({class}`~pymor.reductors.bt.PRBTReductor`) {cite}`DP84,GA04,HJS94`,
- passivity preserving model reduction via spectral factorization
  ({class}`~pymor.reductors.spectral_factor.SpectralFactorReductor`) {cite}`BU22`.

In this section, we apply all three reductors on our toy example and compare
their performance. All three reductors are described in {cite}`BU22` in more detail.

Note: Currently, the {class}`~pymor.reductors.bt.PRBTReductor` and
{class}`~pymor.reductors.spectral_factor.SpectralFactorReductor` reductors require
the symmetric part of {math}`D` (i.e., the $S$ matrix in the port-Hamiltonian system)
to be nonsingular. The MSD example has a zero {math}`D` matrix. Therefore,
we have to add a small regularization feedthrough term, i.e., we replace $D$ with
$D+\varepsilon I_m$. This is a limitation of the current implementation since the

> numerical solution of the KYP-LMI is obtained by solving a related Riccati

Reviewer 2: Maybe simply write the Riccati equation A^T X E + E^T X A + (C^T − X B)(D + D^T)^{-1}(C − B^T X) = 0 here. Then it is more clear what this small regularization term is for and why it is necessary.

```

The `ricc_pos_lrcf` solver option refers to the solver used for the underlying
Riccati equation relevant for {class}`~pymor.reductors.bt.PRBTReductor` and
{class}`~pymor.reductors.spectral_factor.SpectralFactorReductor`. Possible choices are
`scipy`, `slycot` or `pymess` (if installed). Currently, we recommend `slycot`, since pyMOR's
`pymess` binding does not support all use cases yet, and `scipy` gets into trouble if
the associated Hamiltonian pencil has eigenvalues close to the imaginary axis.

### pH-IRKA

The pH-IRKA reductor {class}`~pymor.reductors.ph.ph_irka.PHIRKAReductor` directly returns
a ROM of type {{ PHLTIModel }}. pH-IRKA works similar to the standard IRKA reductor
{class}`~pymor.reductors.h2.IRKAReductor` but with fewer degrees of freedom to preserve
the port-Hamiltonian structure. In more detail, the IRKA fixed-point iteration is performed,
but the left projection matrix is chosen as $W = QV$, which then automatically yields a
reduced pH system with $\hat{Q} = I_r$.

```{code-cell}
from pymor.reductors.ph.ph_irka import PHIRKAReductor

reductor = PHIRKAReductor(fom)
rom1 = reductor.reduce(10)

> print(f'rom1 is of type {type(rom1)}.')

Reviewer: ```suggestion
print(f'rom1 is of type {type(rom1).__qualname__}.')
```

```

### Positive-real balanced truncation (PRBT)

Positive-real balanced truncation (PRBT) works analogously to the standard balanced truncation
method described in {doc}`tutorial_bt`, but uses positive real controllability
and observability Gramians instead. PRBT preserves passivity but returns a ROM
of type {{ LTIModel }}. Thus, we convert the ROM into a {{ PHLTIModel }} in a
post-processing step. Note that PRBT can be used with any passive {{ LTIModel}} FOM.

```{code-cell}
from pymor.reductors.bt import PRBTReductor

reductor = PRBTReductor(fom)
rom2 = reductor.reduce(10)
rom2 = rom2.with_(solver_options={'ricc_pos_lrcf': 'slycot'})
rom2 = PHLTIModel.from_passive_LTIModel(rom2)

> print(f'rom2 is of type {type(rom2)}.')

Reviewer: ```suggestion
print(f'rom2 is of type {type(rom2).__qualname__}.')
```

spectral factor from the FOM (this is only possible if the system is passive),
which subsequentely is reduced by a reductor specified by the user.
A spectral factor is a standard {{ LTIModel }}, and hence any LTI reduction can be used.
For our example, we use the {class}`~pymor.reductors.h2.IRKAReductor` as the inner reductor.
If the inner reductor returns a stable ROM, passivity is preserved.
The spectral factor method and PRBT are related since the computation of
the optimal spectral factor for model reduction depends on the computation of
the positive-real observability Gramian. The spectral factor method can be used with
any passive {{ LTIModel}} FOM. Again, we convert the ROM of type {{ LTIModel}} into a
{{ PHLTIModel }} in a post-processing step.

```{code-cell}
from pymor.reductors.spectral_factor import SpectralFactorReductor
from pymor.reductors.h2 import IRKAReductor

reductor = SpectralFactorReductor(fom)
rom3 = reductor.reduce(
    lambda spectral_factor, mu : IRKAReductor(spectral_factor, mu).reduce(10)
)
rom3 = rom3.with_(solver_options={'ricc_pos_lrcf': 'slycot'})
rom3 = PHLTIModel.from_passive_LTIModel(rom3)

> print(f'rom3 is of type {type(rom3)}.')

Reviewer: ```suggestion
print(f'rom3 is of type {type(rom3).__qualname__}.')
```
